pypesto/model_selection/model_selection.py

```diff
@@ -0,0 +1,906 @@
+        ------
+        Models, one model at a time, as a dictionary, where the keys are the
+        column headers in the model specification file, and the values are the
+        respective column values in a row of the model specification file.
+        """
+        # Go to the start of model specification rows, after the header.
+        self.specification_file.seek(0)
+        self.specification_file.readline()
+
+        if exclusions is None:
+            exclusions = []
+        if exclude_history:
+            exclusions += self.selection_history.keys()
+
+        for line in self.specification_file:
+            model_dict = dict(zip(self.header, line2row(line)))
+            # Exclusion of history makes sense here, to avoid duplicated code
+            # in specific selectors. However, the selection history of this
+            # class is only updated when a `selector.__call__` returns, so
+            # the exclusion of a model tested twice within the same selector
+            # call is not excluded here. Could be implemented by decorating
+            # `model_generator` in `ModelSelectorMethod` to include the
+            # call selection history as `exclusions` (TODO).
+            if model_dict[MODEL_ID] in exclusions:
+                continue
+            yield model_dict
+
+    def select(self, method: str, criterion: str):
+        """
+        Runs a model selection algorithm. The result is the selected model for
+        the current run, independent of previous `select()` calls.
+
+        Arguments
+        ---------
+        method:
+            The model selection algorithm.
+
+        criterion:
+            The criterion used by `ModelSelectorMethod.compare()`, in which
+            currently implemented criterion can be found.
+        """
+        if method == 'forward':
+            selector = ForwardSelector(self.petab_problem,
+                                       self.model_generator,
+                                       criterion,
+                                       self.parameter_IDs,
+                                       self.selection_history)
+            result, self.selection_history = selector()
+        elif method == 'backward':
+            selector = ForwardSelector(self.petab_problem,
+                                       self.model_generator,
+                                       criterion,
+                                       self.parameter_IDs,
+                                       self.selection_history,
+                                       reverse=True)
+            result, self.selection_history = selector()
+        else:
+            raise NotImplementedError(f'Model selection algorithm: {method}.')
+
+        # TODO: Reconsider return value. `result` could be stored in attribute,
+        # then no values need to be returned, and users can request values
+        # manually.
+        return result, self.selection_history
+
+
+class ModelSelectorMethod:
```

Comment:
This is an abstract base class, right? I.e. not intended to be used on its own. Then write `class ModelSelectorMethod(abc.ABC)` and indicate all abstract methods that subclasses have to overwrite by te decorator `@abc.abstractmethod`.